File: engine/static/input/src/cabi/controller.cpp
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <algorithm>
#include "argus/input/cabi/controller.h"
#include "argus/input/controller.hpp"

using argus::input::Controller;

static Controller &_as_ref(argus_controller_t controller) {
    return *reinterpret_cast<Controller *>(controller);
}

static const Controller &_as_ref(argus_controller_const_t controller) {
    return *reinterpret_cast<const Controller *>(controller);
}

extern"C" {

const char *argus_controller_get_name(argus_controller_const_t controller) {
    return _as_ref(controller).get_name().c_str();
}

bool argus_controller_has_gamepad(argus_controller_const_t controller) {
    return _as_ref(controller).has_gamepad();
}

void argus_controller_attach_gamepad(argus_controller_t controller, ArgusHidDeviceId id) {
    _as_ref(controller).attach_gamepad(id);
}

bool argus_controller_attach_first_available_gamepad(argus_controller_t controller) {
    return _as_ref(controller).attach_first_available_gamepad();
}

void argus_controller_detach_gamepad(argus_controller_t controller) {
    _as_ref(controller).detach_gamepad();
}

const char *argus_controller_get_gamepad_name(argus_controller_const_t controller) {
    return _as_ref(controller).get_gamepad_name();
}

double argus_controller_get_deadzone_radius(argus_controller_const_t controller) {
    return _as_ref(controller).get_deadzone_radius();
}

void argus_controller_set_deadzone_radius(argus_controller_t controller, double radius) {
    _as_ref(controller).set_deadzone_radius(radius);
}

void argus_controller_clear_deadzone_radius(argus_controller_t controller) {
    _as_ref(controller).clear_deadzone_radius();
}

ArgusDeadzoneShape argus_controller_get_deadzone_shape(argus_controller_const_t controller) {
    return ArgusDeadzoneShape(_as_ref(controller).get_deadzone_shape());
}

void argus_controller_set_deadzone_shape(argus_controller_t controller, ArgusDeadzoneShape shape) {
    _as_ref(controller).set_deadzone_shape(argus::input::DeadzoneShape(shape));
}

void argus_controller_clear_deadzone_shape(argus_controller_t controller) {
    _as_ref(controller).clear_deadzone_shape();
}

double argus_controller_get_axis_deadzone_radius(argus_controller_const_t controller, ArgusGamepadAxis axis) {
    return _as_ref(controller).get_axis_deadzone_radius(argus::input::GamepadAxis(axis));
}

void argus_controller_set_axis_deadzone_radius(argus_controller_t controller, ArgusGamepadAxis axis, double radius) {
    _as_ref(controller).set_axis_deadzone_radius(argus::input::GamepadAxis(axis), radius);
}

void argus_controller_clear_axis_deadzone_radius(argus_controller_t controller, ArgusGamepadAxis axis) {
    _as_ref(controller).clear_axis_deadzone_radius(argus::input::GamepadAxis(axis));
}

ArgusDeadzoneShape argus_controller_get_axis_deadzone_shape(argus_controller_const_t controller,
        ArgusGamepadAxis axis) {
    return ArgusDeadzoneShape(_as_ref(controller).get_axis_deadzone_shape(argus::input::GamepadAxis(axis)));
}

void argus_controller_set_axis_deadzone_shape(argus_controller_t controller, ArgusGamepadAxis axis,
        ArgusDeadzoneShape shape) {
    _as_ref(controller).set_axis_deadzone_shape(argus::input::GamepadAxis(axis), argus::input::DeadzoneShape(shape));
}

void argus_controller_clear_axis_deadzone_shape(argus_controller_t controller, ArgusGamepadAxis axis) {
    _as_ref(controller).clear_axis_deadzone_shape(argus::input::GamepadAxis(axis));
}

void argus_controller_unbind_action(argus_controller_t controller, const char *action) {
    _as_ref(controller).unbind_action(action);
}

size_t argus_controller_get_keyboard_key_bindings_count(argus_controller_const_t controller,
        ArgusKeyboardScancode key) {
    return _as_ref(controller).get_keyboard_key_bindings(argus::input::KeyboardScancode(key)).size();
}

void argus_controller_get_keyboard_key_bindings(argus_controller_const_t controller, ArgusKeyboardScancode key,
        const char **out_bindings, size_t count) {
    const auto bindings = _as_ref(controller).get_keyboard_key_bindings(argus::input::KeyboardScancode(key));
    affirm_precond(count == bindings.size(),
            "argus_controller_get_keyboard_key_bindings called with wrong count parameter");
    std::transform(bindings.cbegin(), bindings.cend(), out_bindings,
            [](auto action_ref) { return action_ref.get().c_str(); });
}

size_t argus_controller_get_keyboard_action_bindings_count(argus_controller_const_t controller, const char *action) {
    return _as_ref(controller).get_keyboard_action_bindings(action).size();
}

void argus_controller_get_keyboard_action_bindings(argus_controller_const_t controller, const char *action,
        ArgusKeyboardScancode *out_scancodes, size_t count) {
    const auto bindings = _as_ref(controller).get_keyboard_action_bindings(action);
    affirm_precond(count == bindings.size(),
            "argus_controller_get_keyboard_action_bindings called with wrong count parameter");
    std::transform(bindings.cbegin(), bindings.cend(), out_scancodes,
            [](auto key) { return ArgusKeyboardScancode(key); });
}

void argus_controller_bind_keyboard_key(argus_controller_t controller, ArgusKeyboardScancode key, const char *action) {
    _as_ref(controller).bind_keyboard_key(argus::input::KeyboardScancode(key), action);
}

void argus_controller_unbind_keyboard_key(argus_controller_t controller, ArgusKeyboardScancode key) {
    _as_ref(controller).unbind_keyboard_key(argus::input::KeyboardScancode(key));
}

void argus_controller_unbind_keyboard_key_action(argus_controller_t controller, ArgusKeyboardScancode key,
        const char *action) {
    _as_ref(controller).unbind_keyboard_key(argus::input::KeyboardScancode(key), action);
}

void argus_controller_bind_mouse_button(argus_controller_t controller, ArgusMouseButton button, const char *action) {
    _as_ref(controller).bind_mouse_button(argus::input::MouseButton(button), action);
}

void argus_controller_unbind_mouse_button(argus_controller_t controller, ArgusMouseButton button) {
    _as_ref(controller).unbind_mouse_button(argus::input::MouseButton(button));
}

void argus_controller_unbind_mouse_button_action(argus_controller_t controller, ArgusMouseButton button,
        const char *action) {
    _as_ref(controller).unbind_mouse_button(argus::input::MouseButton(button), action);
}

void argus_controller_bind_mouse_axis(argus_controller_t controller, ArgusMouseAxis axis, const char *action) {
    _as_ref(controller).bind_mouse_axis(argus::input::MouseAxis(axis), action);
}

void argus_controller_unbind_mouse_axis(argus_controller_t controller, ArgusMouseAxis axis) {
    _as_ref(controller).unbind_mouse_axis(argus::input::MouseAxis(axis));
}

void argus_controller_unbind_mouse_axis_action(argus_controller_t controller, ArgusMouseAxis axis, const char *action) {
    _as_ref(controller).unbind_mouse_axis(argus::input::MouseAxis(axis), action);
}

void argus_controller_bind_gamepad_button(argus_controller_t controller, ArgusGamepadButton button,
        const char *action) {
    _as_ref(controller).bind_gamepad_button(argus::input::GamepadButton(button), action);
}

void argus_controller_unbind_gamepad_button(argus_controller_t controller, ArgusGamepadButton button) {
    _as_ref(controller).unbind_gamepad_button(argus::input::GamepadButton(button));
}

void argus_controller_unbind_gamepad_button_action(argus_controller_t controller, ArgusGamepadButton button,
        const char *action) {
    _as_ref(controller).unbind_gamepad_button(argus::input::GamepadButton(button), action);
}

void argus_controller_bind_gamepad_axis(argus_controller_t controller, ArgusGamepadAxis axis, const char *action) {
    _as_ref(controller).bind_gamepad_axis(argus::input::GamepadAxis(axis), action);
}

void argus_controller_unbind_gamepad_axis(argus_controller_t controller, ArgusGamepadAxis axis) {
    _as_ref(controller).unbind_gamepad_axis(argus::input::GamepadAxis(axis));
}

void argus_controller_unbind_gamepad_axis_action(argus_controller_t controller, ArgusGamepadAxis axis,
        const char *action) {
    _as_ref(controller).unbind_gamepad_axis(argus::input::GamepadAxis(axis), action);
}

bool argus_controller_is_gamepad_button_pressed(argus_controller_const_t controller, ArgusGamepadButton button) {
    return _as_ref(controller).is_gamepad_button_pressed(argus::input::GamepadButton(button));
}

double argus_controller_get_gamepad_axis(argus_controller_const_t controller, ArgusGamepadAxis axis) {
    return _as_ref(controller).get_gamepad_axis(argus::input::GamepadAxis(axis));
}

double argus_controller_get_gamepad_axis_delta(argus_controller_const_t controller, ArgusGamepadAxis axis) {
    return _as_ref(controller).get_gamepad_axis_delta(argus::input::GamepadAxis(axis));
}

bool argus_controller_is_action_pressed(argus_controller_const_t controller, const char *action) {
    return _as_ref(controller).is_action_pressed(action);
}

double argus_controller_get_action_axis(argus_controller_const_t controller, const char *action) {
    return _as_ref(controller).get_action_axis(action);
}

double argus_controller_get_action_axis_delta(argus_controller_const_t controller, const char *action) {
    return _as_ref(controller).get_action_axis_delta(action);
}

}

File: engine/auxiliary/input_rustabi/src/argus/input/controller.rs
/*
 * This file is a part of Argus.
 * Copyright (c) 2019-2024, Max Roncace <mproncace@protonmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
use std::{ffi, ptr};
use std::ffi::{c_void, CStr, CString};
use num_enum::{IntoPrimitive, UnsafeFromPrimitive};
use lowlevel_rustabi::argus::lowlevel::FfiWrapper;
use crate::argus::input::{GamepadAxis, GamepadButton, HidDeviceId, KeyboardScancode, MouseAxis, MouseButton};
use crate::input_cabi::*;

#[repr(u32)]
#[derive(Clone, Debug, IntoPrimitive, UnsafeFromPrimitive)]
pub enum DeadzoneShape {
    Ellipse = DEADZONE_SHAPE_ELLIPSE,
    Quad = DEADZONE_SHAPE_QUAD,
    Cross = DEADZONE_SHAPE_CROSS,
    MaxValue = DEADZONE_SHAPE_MAXVALUE,
}

pub struct Controller {
    handle: argus_controller_t,
}

impl FfiWrapper for Controller {
    fn of(handle: *mut c_void) -> Self {
        Self { handle }
    }
}

impl Controller {
    pub fn get_name(&self) -> String {
        unsafe {
            CStr::from_ptr(argus_controller_get_name(self.handle)).to_string_lossy().to_string()
        }
    }

    pub fn has_gamepad(&self) -> bool {
        unsafe { argus_controller_has_gamepad(self.handle) }
    }

    pub fn attach_gamepad(&mut self, id: HidDeviceId) {
        unsafe { argus_controller_attach_gamepad(self.handle, id); }
    }

    pub fn attach_first_available_gamepad(&mut self) -> bool {
        unsafe { argus_controller_attach_first_available_gamepad(self.handle) }
    }

    pub fn detach_gamepad(&mut self) {
        unsafe { argus_controller_detach_gamepad(self.handle); }
    }

    pub fn get_gamepad_name(&self) -> String {
        unsafe {
            CStr::from_ptr(argus_controller_get_gamepad_name(self.handle))
                .to_string_lossy().to_string()
        }
    }

    pub fn get_deadzone_radius(&self) -> f64 {
        unsafe { argus_controller_get_deadzone_radius(self.handle) }
    }

    pub fn set_deadzone_radius(&mut self, radius: f64) {
        unsafe { argus_controller_set_deadzone_radius(self.handle, radius); }
    }

    pub fn clear_deadzone_radius(&mut self) {
        unsafe { argus_controller_clear_deadzone_radius(self.handle); }
    }

    pub fn get_deadzone_shape(&self) -> DeadzoneShape {
        unsafe {
            DeadzoneShape::unchecked_transmute_from(
                argus_controller_get_deadzone_shape(self.handle)
            )
        }
    }

    pub fn set_deadzone_shape(&mut self, shape: DeadzoneShape) {
        unsafe { argus_controller_set_deadzone_shape(self.handle, shape.into()); }
    }

    pub fn clear_deadzone_shape(&mut self) {
        unsafe { argus_controller_clear_deadzone_shape(self.handle) }
    }

    pub fn get_axis_deadzone_radius(&self, axis: GamepadAxis) -> f64 {
        unsafe { argus_controller_get_axis_deadzone_radius(self.handle, axis.into()) }
    }

    pub fn set_axis_deadzone_radius(&mut self, axis: GamepadAxis, radius: f64) {
        unsafe { argus_controller_set_axis_deadzone_radius(self.handle, axis.into(), radius); }
    }

    pub fn clear_axis_deadzone_radius(&mut self, axis: GamepadAxis) {
        unsafe { argus_controller_clear_axis_deadzone_radius(self.handle, axis.into()); }
    }

    pub fn get_axis_deadzone_shape(&self, axis: GamepadAxis) -> DeadzoneShape {
        unsafe {
            DeadzoneShape::unchecked_transmute_from(
                argus_controller_get_axis_deadzone_shape(self.handle, axis.into())
            )
        }
    }

    pub fn set_axis_deadzone_shape(&mut self, axis: GamepadAxis, shape: DeadzoneShape) {
        unsafe { argus_controller_set_axis_deadzone_shape(self.handle, axis.into(), shape.into()); }
    }

    pub fn clear_axis_deadzone_shape(&mut self, axis: GamepadAxis) {
        unsafe { argus_controller_clear_axis_deadzone_shape(self.handle, axis.into()); }
    }

    pub fn unbind_action(&mut self, action: &str) {
        unsafe {
            let action_c = CString::new(action).unwrap();
            argus_controller_unbind_action(self.handle, action_c.as_ptr());
        }
    }

    pub fn get_keyboard_key_bindings(&self, key: KeyboardScancode) -> Vec<String> {
        unsafe {
            let count = argus_controller_get_keyboard_key_bindings_count(
                self.handle,
                key.clone().into(),
            );
            let mut bindings: Vec<*const ffi::c_char> = Vec::with_capacity(count);
            bindings.resize(count, ptr::null());
            argus_controller_get_keyboard_key_bindings(
                self.handle,
                key.into(),
                bindings.as_mut_ptr(),
                count,
            );
            bindings.into_iter()
                .map(|action| CStr::from_ptr(action).to_string_lossy().to_string())
                .collect()
        }
    }

    pub fn get_keyboard_action_bindings(&self, action: &str) -> Vec<KeyboardScancode> {
        let action_c = CString::new(action).unwrap();
        unsafe {
            let count = argus_controller_get_keyboard_action_bindings_count(
                self.handle,
                action_c.as_ptr(),
            );
            let mut bindings: Vec<KeyboardScancode> = Vec::with_capacity(count);
            bindings.resize(count, KeyboardScancode::Unknown);
            argus_controller_get_keyboard_action_bindings(
                self.handle,
                action_c.as_ptr(),
                bindings.as_mut_ptr().cast(),
                count,
            );
            bindings
        }
    }

    pub fn bind_keyboard_key(&mut self, key: KeyboardScancode, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_bind_keyboard_key(self.handle, key.into(), action_c.as_ptr()); }
    }

    pub fn unbind_keyboard_key(&mut self, key: KeyboardScancode) {
        unsafe { argus_controller_unbind_keyboard_key(self.handle, key.into()); }
    }

    pub fn unbind_keyboard_key_action(&mut self, key: KeyboardScancode, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_unbind_keyboard_key_action(self.handle, key.into(), action_c.as_ptr());
        }
    }

    pub fn bind_mouse_button(&mut self, button: MouseButton, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_bind_mouse_button(self.handle, button.into(), action_c.as_ptr());
        }
    }

    pub fn unbind_mouse_button(&mut self, button: MouseButton) {
        unsafe { argus_controller_unbind_mouse_button(self.handle, button.into()); }
    }

    pub fn unbind_mouse_button_action(&mut self, button: MouseButton, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_unbind_mouse_button_action(
                self.handle,
                button.into(),
                action_c.as_ptr(),
            );
        }
    }

    pub fn bind_mouse_axis(&mut self, axis: MouseAxis, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_bind_mouse_axis(self.handle, axis.into(), action_c.as_ptr()); }
    }

    pub fn unbind_mouse_axis(&mut self, axis: MouseAxis) {
        unsafe { argus_controller_unbind_mouse_axis(self.handle, axis.into()); }
    }

    pub fn unbind_mouse_axis_action(&mut self, axis: MouseAxis, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_unbind_mouse_axis_action(self.handle, axis.into(), action_c.as_ptr());
        }
    }

    pub fn bind_gamepad_button(&mut self, button: GamepadButton, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_bind_gamepad_button(self.handle, button.into(), action_c.as_ptr());
        }
    }

    pub fn unbind_gamepad_button(&mut self, button: GamepadButton) {
        unsafe { argus_controller_unbind_gamepad_button(self.handle, button.into()); }
    }

    pub fn unbind_gamepad_button_action(&mut self, button: GamepadButton, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_unbind_gamepad_button_action(
                self.handle,
                button.into(),
                action_c.as_ptr(),
            );
        }
    }

    pub fn bind_gamepad_axis(&mut self, axis: GamepadAxis, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_bind_gamepad_axis(self.handle, axis.into(), action_c.as_ptr()); }
    }

    pub fn unbind_gamepad_axis(&mut self, axis: GamepadAxis) {
        unsafe { argus_controller_unbind_gamepad_axis(self.handle, axis.into()); }
    }

    pub fn unbind_gamepad_axis_action(&mut self, axis: GamepadAxis, action: &str) {
        let action_c = CString::new(action).unwrap();
        unsafe {
            argus_controller_unbind_gamepad_axis_action(
                self.handle,
                axis.into(),
                action_c.as_ptr()
            );
        }
    }

    pub fn is_gamepad_button_pressed(&self, button: GamepadButton) -> bool {
        unsafe { argus_controller_is_gamepad_button_pressed(self.handle, button.into()) }
    }

    pub fn get_gamepad_axis(&self, axis: GamepadAxis) -> f64 {
        unsafe { argus_controller_get_gamepad_axis(self.handle, axis.into()) }
    }

    pub fn get_gamepad_axis_delta(&self, axis: GamepadAxis) -> f64 {
        unsafe { argus_controller_get_gamepad_axis_delta(self.handle, axis.into()) }
    }

    pub fn is_action_pressed(&self, action: &str) -> bool {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_is_action_pressed(self.handle, action_c.as_ptr()) }
    }

    pub fn get_action_axis(&self, action: &str) -> f64 {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_get_action_axis(self.handle, action_c.as_ptr()) }
    }

    pub fn get_action_axis_delta(&self, action: &str) -> f64 {
        let action_c = CString::new(action).unwrap();
        unsafe { argus_controller_get_action_axis_delta(self.handle, action_c.as_ptr()) }
    }
}
